drgn/helpers/linux/dmesg.py

```diff
@@ -0,0 +1,166 @@
+
+"""
+dmesg-vmcore
+-------
+
+The ``drgn.helpers.linux.dmesg`` module provides helpers for reading the
+Linux kernel log.
+"""
+
+from typing import Dict, List, NamedTuple
+
+from drgn import Object, Program, cast, sizeof
+
+__all__ = ("get_dmesg", "get_printk_records")
+
+
+class PrintkRecord(NamedTuple):
+    text: bytes
+    facility: int
+    level: int
+    seq: int
+    timestamp: int
+    continuation: bool
+    context: Dict[bytes, bytes]
+
+
+def get_printk_records(prog: Program) -> List[PrintkRecord]:
+    try:
+        prb = prog["prb"]
+    except KeyError:
+        return dump_dmesg_structured(prog)
+    else:
+        return dump_dmesg_lockless(prog, prb)
+
+
+def dump_dmesg_lockless(prog: Program, printk_ringbuffer: Object) -> List[PrintkRecord]:
+    DESC_SV_BITS = sizeof(prog.type("unsigned long")) * 8
+    DESC_FLAGS_SHIFT = DESC_SV_BITS - 2
+    DESC_FLAGS_MASK = 3 << DESC_FLAGS_SHIFT
+    DESC_ID_MASK = ~DESC_FLAGS_MASK
+
+    desc_ring = printk_ringbuffer.desc_ring
+    desc_ring_count = 1 << desc_ring.count_bits
+    text_data_ring = printk_ringbuffer.text_data_ring
+    text_data_ring_size = 1 << text_data_ring.size_bits
+
+    tail_id = desc_ring.tail_id.counter
+    head_id = desc_ring.head_id.counter
+    current_id = tail_id
+
+    out_buf = str.encode("")
+    ulong_size = prog.type("long").size
+
+    desc_committed = prog["desc_committed"]
+    desc_finalized = prog["desc_finalized"]
+
+    def id_incr(counter: Object) -> Object:
+        return (counter + 1) & DESC_ID_MASK
+
+    def record_committed(current_id: Object, state_var: Object) -> bool:
+        state_desc_id = state_var & DESC_ID_MASK
+        state = 3 & (state_var >> DESC_FLAGS_SHIFT)
+        return (current_id == state_desc_id) and (
+            state == desc_committed or state == desc_finalized
+        )
+
+    def dump_record(current_id: Object) -> List[PrintkRecord]:
+        idx = current_id % desc_ring_count
+        descs = desc_ring.descs[idx]
+        infos = desc_ring.infos[idx]
+        text_len = infos.text_len
+        facility = infos.facility.value_()
+        level = infos.level.value_()
+        seq = infos.seq.value_()
+        timestamp = infos.ts_nsec.value_()
+        context = {
+            bytes("SUBSYSTEM", "utf-8"): bytes(infos.dev_info.subsystem),
+            bytes("DEVICE", "utf-8"): bytes(infos.dev_info.device),
+        }
+
+        if not record_committed(current_id, descs.state_var.counter):
+            return []
+
+        lpos_begin = descs.text_blk_lpos.begin % text_data_ring_size
+        lpos_next = descs.text_blk_lpos.next % text_data_ring_size
+        lpos_begin += ulong_size
+
+        if lpos_begin == lpos_next:
+            return []
+        if lpos_begin > lpos_next:
+            lpos_begin -= lpos_begin
+        if lpos_next - lpos_begin < text_len:
+            text_len = lpos_next - lpos_begin
+
+        text = prog.read(text_data_ring.data + lpos_begin, text_len) + str.encode("\n")
```

Comment:
I chose to take the newline out of the text, and join on newline in get_dmesg. Otherwise it's rather awkward to have to add a newline to the text section when we read from /dev/kmsg in the test.